pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,455 @@
+				select {
+				case <-q.notifyStop:
+					return
+				case <-q.notifyNoMoreItemsComplete:
+					break
+				}
+				break
+			}
+		} else if dwWait == (windows.WAIT_OBJECT_0 + 1) {
+			// Stop event is set
+			return
+		} else if dwWait == uint32(windows.WAIT_TIMEOUT) {
+			// timeout
+			// this shouldn't happen
+			pkglog.Errorf("WaitForMultipleObjects timed out")
+			return
+		} else {
+			// some other error occurred
+			gle := windows.GetLastError()
+			pkglog.Errorf("WaitForMultipleObjects unknown error: wait(%d,%#x) gle(%d,%#x)",
+				dwWait,
+				dwWait,
+				gle,
+				gle)
+			return
+		}
+	}
+}
+
+// synchronizeNoMoreItems is used to synchronize notifyEventsAvailableLoop when
+// EvtNext returns ERROR_NO_MORE_ITEMS.
+// Note that the Microsoft's Pull Subscription model is inherently racey. It is possible
+// for EvtNext to return ERROR_NO_MORE_ITEMS and then for Windows to call SetEvent(waitHandle)
+// before our code reaches the ResetEvent(waitHandle). If this happens we will not see those
+// events until newer events are created and Windows once again calls SetEvent(waitHandle).
+func (q *pullSubscription) synchronizeNoMoreItems() error {
+	if !q.useNotifyChannel {
+		_ = windows.ResetEvent(windows.Handle(q.waitEventHandle))
+		return nil
+	}
+
+	// If notifyEventsAvailableLoop is blocking on WaitForMultipleObjects
+	// wake it up so we can sync on notifyNoMoreItems
+	// If notifyEventsAvailableLoop is blocking on notifyNoMoreItems channel then this is a no-op
+	windows.SetEvent(windows.Handle(q.waitEventHandle))
+	// If notifyEventsAvailableLoop is blocking on sending notifyEventsAvailable
+	// then wake/cancel it so it does not erroneously send notifyEventsAvailable.
+	select {
+	case <-q.notifyStop:
+		return fmt.Errorf("stop signal")
+	case <-q.notifyEventsAvailableLoopDone:
+		return fmt.Errorf("notify loop is not running")
+	case q.notifyNoMoreItems <- struct{}{}:
+		break
+	}
+	// Reset the events ready event so notifyEventsAvailableLoop will wait again in WaitForMultipleObjects,
+	// then write to notifyNoMoreItemsComplete to tell the loop that the event has been reset and it
+	// can safely continue.
+	_ = windows.ResetEvent(windows.Handle(q.waitEventHandle))
+	select {
+	case <-q.notifyStop:
+		return fmt.Errorf("stop signal")
+	case <-q.notifyEventsAvailableLoopDone:
+		return fmt.Errorf("notify loop is not running")
+	case q.notifyNoMoreItemsComplete <- struct{}{}:
+		break
+	}
+	return nil
+}
+
+// GetEvents returns the next available events in the subscription.
+func (q *pullSubscription) GetEvents() ([]*evtapi.EventRecord, error) {
+	// Do a non-blocking check to see if the "events available" event handle is set.
+	// We should only call EvtNext when this event is set.
+	dwWait, err := windows.WaitForSingleObject(windows.Handle(q.waitEventHandle), 0)
+	if err != nil {
+		return nil, fmt.Errorf("WaitForSingleObject failed: %v", err)
+	}
+	if dwWait != windows.WAIT_OBJECT_0 {
```

Comment:
This is here so the `PullSubscription` can be used without the go notify channel, in a polling scenario rather than a wait scenario. Which fits better into checks which run and check for events on an interval.

Also since `EvtNext` will return an error if you do call it in this state, it helps avoid the user from thinking there is a fatal error. I updated the comment.

This is tested in `TestUnusedNotifyChannel`

Reply:
there was a tight loop on this error in the tailer though. fixed by https://github.com/DataDog/datadog-agent/pull/19761/commits/655b083db288e7c104c81786141724ececacd11f